### PR TITLE
Hack week 2

### PR DESCRIPTION
**Pull request by yasinzaehringer-paradime**
support for:
- `{{ "true" if False }}` and `{{ "true" if True else 'false' }}`
- `{{ a.b[c.d] }}`

bugfixes for (found using the fuzzer):
- `{{ {'dict': 'of', 'key': 'and', 'value': 'p`
- `{%for A in 0%}{%endfor`
- `{{0 is`
- `{{ "a"`

### CODE REVIEW

File: testData/fuzz/incomplete-dictionary.err.out [hunk 1]
```diff
@@ -0,0 +1 @@
+Expected either a number, string, keyword or identifier.*
```
Comment (bhuvansingla) [inline testData/fuzz/incomplete-dictionary.err.out:1]: what does `.*` mean here?

File: testData/fuzz/incomplete-is-stmt.err.out [hunk 2]
```diff
@@ -0,0 +1 @@
+is statement is incomplete
```
Comment (bhuvansingla) [inline testData/fuzz/incomplete-is-stmt.err.out:1]: could be confusing to read at first, but I don't have anything better too.

File: testData/statements/issue-hackweek.tpl [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{% set a={"b":{"key":"value"}} %}
```
Comment (bhuvansingla) [inline testData/statements/issue-hackweek.tpl]: better filename?

File: builtins/statements/set.go [hunk 4]
```diff
@@ -52,15 +52,19 @@ func (stmt *SetStmt) Execute(r *exec.Renderer, tag *nodes.StatementBlock) error
 		if target.IsError() {
 			return errors.Wrapf(target, `Unable to evaluate target %s`, n)
 		}
-		if err := target.Set(n.Attr, value.Interface()); err != nil {
+		if n.Attr == "" {
+			return errors.Errorf(`Not implemented to evaluate getattr at %s`, n.Attr) // TODO: implement
```
Comment (bhuvansingla) [inline builtins/statements/set.go]: `...%s, n.Attr...` not really useful. n.Attr is always empty

File: builtins/statements/set.go [hunk 5]
```diff
@@ -52,15 +52,19 @@ func (stmt *SetStmt) Execute(r *exec.Renderer, tag *nodes.StatementBlock) error
...
 			return errors.Wrapf(err, `Unable to set value on "%s"`, n.Attr)
 		}
 	case *nodes.Getitem:
 		target := r.Eval(n.Node)
 		if target.IsError() {
 			return errors.Wrapf(target, `Unable to evaluate target %s`, n)
 		}
-		if err := target.Set(n.Arg, value.Interface()); err != nil {
+		if n.Arg == nil {
+			return errors.Errorf(`Not implemented to evaluate getitem at %s`, n.Arg) // TODO: implement
```
Comment (bhuvansingla) [inline builtins/statements/set.go]: not sure if %s works with nil value

File: parser/expression.go [hunk 6]
```diff
@@ -40,9 +40,18 @@ func (p *Parser) ParseFilterExpression(expr nodes.Expression) (nodes.Expression,
 	return expr, nil
 }
 
-// ParseExpression parses an expression with optionnal filters
+// ParseExpression parses an expression with optional filters
 // Nested expression shoulds call this method
+
 func (p *Parser) ParseExpression() (nodes.Expression, error) {
+	return p.parseExpression(false)
+}
```
Comment (bhuvansingla) [inline parser/expression.go:48]: Do we really need the case with no inline ifs?

**Review by bhuvansingla — APPROVED** [review]
approved, but left few comments

File: testData/fuzz/incomplete-dictionary.err.out (hunk 1 above)
Comment (yasinzaehringer-paradime) [reply testData/fuzz/incomplete-dictionary.err.out:1]: This is a regular expression.

File: parser/expression.go (hunk 6 above)
Comment (yasinzaehringer-paradime) [reply parser/expression.go:48]: yes: luckily there were test cases which caught that: jinja templates support this pattern:
```
{%- for person in persons if person.Gender is equalto "male" %}
{{ person.FirstName }} {{ person.LastName }}
{%- endfor %}
```
Source: https://github.com/paradime-io/gonja/blob/master/testData/statements/for.tpl#L46

On the other hand you are right: one can further improve the jinja compatibility.

File: builtins/statements/set.go (hunk 5 above)
Comment (yasinzaehringer-paradime) [reply builtins/statements/set.go]: actual it's printing the wrong thing: https://github.com/paradime-io/gonja/pull/1/commits/5f821f23d18e8f41fd9d76896dac5158cf0977f4

File: builtins/statements/set.go (hunk 4 above)
Comment (yasinzaehringer-paradime) [reply builtins/statements/set.go]: actual it's printing the wrong thing: https://github.com/paradime-io/gonja/pull/1/commits/5f821f23d18e8f41fd9d76896dac5158cf0977f4

File: testData/statements/issue-hackweek.tpl (hunk 3 above)
Comment (yasinzaehringer-paradime) [reply testData/statements/issue-hackweek.tpl]: https://github.com/paradime-io/gonja/pull/1/commits/34f6734a8cabc325219b68e54b6b09f5f213f91b

File: builtins/statements/set.go (hunk 4 above)
Comment (yasinzaehringer-paradime) [reply builtins/statements/set.go]: well spotted!

**Pull request merged**